### ISSUE TITLE
XMLStreamException - Premature end of file

### ISSUE THREAD

**Issue by mchlrch** [carml/carml#315]
When attempting to convert XML files, we frequently run into `XMLStreamException: ParseError at [row,col]:[1,1]
Message: Premature end of file.`.

With same code and same input source and mapping this happens frequently but not always.

I created a sample project that works for me to provoke the exception (using carml version 0.4.8): 
https://github.com/mchlrch/carml-xml-issue-1

Be aware that it doesn't reliably reproduce the problem, so sometimes `XmlTestFail` may succeed and `XmlTestSuccess` may fail. It is strange and random, but if I toggle between them and try couple of times, then sooner or later the exception shows up. 

It also seems that some parts are static. So although we build a new `RdfRmlMapper` for each conversion, once the exception occurs, subsequent attempts in the same JVM will fail. I also noticed, that it doesn't work first and then fails, but fails on the first conversion attempt.

Steps to reproduce:

```
mira@mauna:~/git/carml-xml-issue-1$ java --version
openjdk 11.0.15 2022-04-19
OpenJDK Runtime Environment (build 11.0.15+10-Ubuntu-0ubuntu0.20.04.1)
OpenJDK 64-Bit Server VM (build 11.0.15+10-Ubuntu-0ubuntu0.20.04.1, mixed mode, sharing)

mira@mauna:~/git/carml-xml-issue-1$ mvn exec:java -Dexec.mainClass="XmlTestSuccess"
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/usr/share/maven/lib/guice.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[INFO] Scanning for projects...
[INFO]
[INFO] --------------------< com.zazuko:carml-xml-issue-1 >--------------------
[INFO] Building carml-xml-issue-1 0.0.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- exec-maven-plugin:3.1.0:java (default-cli) @ carml-xml-issue-1 ---
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.408 s
[INFO] Finished at: 2023-09-06T14:26:33+02:00
[INFO] ------------------------------------------------------------------------

mira@mauna:~/git/carml-xml-issue-1$ mvn exec:java -Dexec.mainClass="XmlTestFail"
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/usr/share/maven/lib/guice.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[INFO] Scanning for projects...
[INFO]
[INFO] --------------------< com.zazuko:carml-xml-issue-1 >--------------------
[INFO] Building carml-xml-issue-1 0.0.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- exec-maven-plugin:3.1.0:java (default-cli) @ carml-xml-issue-1 ---
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.
[WARNING]
io.carml.logicalsourceresolver.LogicalSourceResolverException: Error executing XPath expression.
    at io.carml.logicalsourceresolver.XPathResolver.xpathPathFlux (XPathResolver.java:154)
    at io.carml.logicalsourceresolver.XPathResolver.lambda$getXpathResultFlux$2 (XPathResolver.java:120)
    at reactor.core.publisher.FluxCreate.subscribe (FluxCreate.java:95)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap (FluxFlatMap.java:200)
    at reactor.core.publisher.FluxFlatMap.subscribeOrReturn (FluxFlatMap.java:93)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8657)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext (FluxFlatMap.java:427)
    at reactor.core.publisher.FluxIterable$IterableSubscription.slowPath (FluxIterable.java:335)
    at reactor.core.publisher.FluxIterable$IterableSubscription.request (FluxIterable.java:294)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onSubscribe (FluxFlatMap.java:371)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:201)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:83)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete (FluxConcatArray.java:258)
    at reactor.core.publisher.FluxConcatArray.subscribe (FluxConcatArray.java:78)
    at reactor.core.publisher.Mono.subscribe (Mono.java:4485)
    at reactor.core.publisher.Mono.block (Mono.java:1733)
    at io.carml.engine.rdf.RdfRmlMapper.toModel (RdfRmlMapper.java:368)
    at io.carml.engine.rdf.RdfRmlMapper.mapToModel (RdfRmlMapper.java:351)
    at XmlTestCommon.doConversion_0 (XmlTestCommon.java:57)
    at XmlTestFail.main (XmlTestFail.java:6)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:279)
    at java.lang.Thread.run (Thread.java:829)
Caused by: javax.xml.xpath.XPathException: org.xml.sax.SAXException: javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]
Message: Premature end of file.
javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]
Message: Premature end of file.
    at jlibs.xml.sax.dog.XMLDog.sniff (XMLDog.java:192)
    at io.carml.logicalsourceresolver.XPathResolver.xpathPathFlux (XPathResolver.java:152)
    at io.carml.logicalsourceresolver.XPathResolver.lambda$getXpathResultFlux$2 (XPathResolver.java:120)
    at reactor.core.publisher.FluxCreate.subscribe (FluxCreate.java:95)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap (FluxFlatMap.java:200)
    at reactor.core.publisher.FluxFlatMap.subscribeOrReturn (FluxFlatMap.java:93)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8657)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext (FluxFlatMap.java:427)
    at reactor.core.publisher.FluxIterable$IterableSubscription.slowPath (FluxIterable.java:335)
    at reactor.core.publisher.FluxIterable$IterableSubscription.request (FluxIterable.java:294)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onSubscribe (FluxFlatMap.java:371)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:201)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:83)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete (FluxConcatArray.java:258)
    at reactor.core.publisher.FluxConcatArray.subscribe (FluxConcatArray.java:78)
    at reactor.core.publisher.Mono.subscribe (Mono.java:4485)
    at reactor.core.publisher.Mono.block (Mono.java:1733)
    at io.carml.engine.rdf.RdfRmlMapper.toModel (RdfRmlMapper.java:368)
    at io.carml.engine.rdf.RdfRmlMapper.mapToModel (RdfRmlMapper.java:351)
    at XmlTestCommon.doConversion_0 (XmlTestCommon.java:57)
    at XmlTestFail.main (XmlTestFail.java:6)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:279)
    at java.lang.Thread.run (Thread.java:829)
Caused by: org.xml.sax.SAXException: javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]
Message: Premature end of file.
    at io.carml.logicalsourceresolver.PausableStaxXmlReader.run (PausableStaxXmlReader.java:152)
    at io.carml.logicalsourceresolver.PausableStaxXmlReader.start (PausableStaxXmlReader.java:103)
    at io.carml.logicalsourceresolver.PausableStaxXmlReader.parse (PausableStaxXmlReader.java:75)
    at jlibs.xml.sax.dog.XMLDog.sniff (XMLDog.java:189)
    at io.carml.logicalsourceresolver.XPathResolver.xpathPathFlux (XPathResolver.java:152)
    at io.carml.logicalsourceresolver.XPathResolver.lambda$getXpathResultFlux$2 (XPathResolver.java:120)
    at reactor.core.publisher.FluxCreate.subscribe (FluxCreate.java:95)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap (FluxFlatMap.java:200)
    at reactor.core.publisher.FluxFlatMap.subscribeOrReturn (FluxFlatMap.java:93)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8657)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext (FluxFlatMap.java:427)
    at reactor.core.publisher.FluxIterable$IterableSubscription.slowPath (FluxIterable.java:335)
    at reactor.core.publisher.FluxIterable$IterableSubscription.request (FluxIterable.java:294)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onSubscribe (FluxFlatMap.java:371)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:201)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:83)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete (FluxConcatArray.java:258)
    at reactor.core.publisher.FluxConcatArray.subscribe (FluxConcatArray.java:78)
    at reactor.core.publisher.Mono.subscribe (Mono.java:4485)
    at reactor.core.publisher.Mono.block (Mono.java:1733)
    at io.carml.engine.rdf.RdfRmlMapper.toModel (RdfRmlMapper.java:368)
    at io.carml.engine.rdf.RdfRmlMapper.mapToModel (RdfRmlMapper.java:351)
    at XmlTestCommon.doConversion_0 (XmlTestCommon.java:57)
    at XmlTestFail.main (XmlTestFail.java:6)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:279)
    at java.lang.Thread.run (Thread.java:829)
Caused by: javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]
Message: Premature end of file.
    at com.sun.org.apache.xerces.internal.impl.XMLStreamReaderImpl.next (XMLStreamReaderImpl.java:652)
    at io.carml.logicalsourceresolver.PausableStaxXmlReader.run (PausableStaxXmlReader.java:150)
    at io.carml.logicalsourceresolver.PausableStaxXmlReader.start (PausableStaxXmlReader.java:103)
    at io.carml.logicalsourceresolver.PausableStaxXmlReader.parse (PausableStaxXmlReader.java:75)
    at jlibs.xml.sax.dog.XMLDog.sniff (XMLDog.java:189)
    at io.carml.logicalsourceresolver.XPathResolver.xpathPathFlux (XPathResolver.java:152)
    at io.carml.logicalsourceresolver.XPathResolver.lambda$getXpathResultFlux$2 (XPathResolver.java:120)
    at reactor.core.publisher.FluxCreate.subscribe (FluxCreate.java:95)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap (FluxFlatMap.java:200)
    at reactor.core.publisher.FluxFlatMap.subscribeOrReturn (FluxFlatMap.java:93)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8657)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext (FluxFlatMap.java:427)
    at reactor.core.publisher.FluxIterable$IterableSubscription.slowPath (FluxIterable.java:335)
    at reactor.core.publisher.FluxIterable$IterableSubscription.request (FluxIterable.java:294)
    at reactor.core.publisher.FluxFlatMap$FlatMapMain.onSubscribe (FluxFlatMap.java:371)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:201)
    at reactor.core.publisher.FluxIterable.subscribe (FluxIterable.java:83)
    at reactor.core.publisher.Flux.subscribe (Flux.java:8671)
    at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onComplete (FluxConcatArray.java:258)
    at reactor.core.publisher.FluxConcatArray.subscribe (FluxConcatArray.java:78)
    at reactor.core.publisher.Mono.subscribe (Mono.java:4485)
    at reactor.core.publisher.Mono.block (Mono.java:1733)
    at io.carml.engine.rdf.RdfRmlMapper.toModel (RdfRmlMapper.java:368)
    at io.carml.engine.rdf.RdfRmlMapper.mapToModel (RdfRmlMapper.java:351)
    at XmlTestCommon.doConversion_0 (XmlTestCommon.java:57)
    at XmlTestFail.main (XmlTestFail.java:6)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:279)
    at java.lang.Thread.run (Thread.java:829)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.544 s
[INFO] Finished at: 2023-09-06T14:26:39+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:3.1.0:java (default-cli) on project carml-xml-issue-1: An exception occurred while executing the Java class. Error executing XPath expression.: org.xml.sax.SAXException: javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]
[ERROR] Message: Premature end of file.
[ERROR] javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,1]
[ERROR] Message: Premature end of file.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

**Comment by pmaria** [comment]
@mchlrch thanks for the reproducer. I'm able to reproduce the issue. Will investigate.

**Comment by pmaria** [comment]
Ok, I found the issue.

The problem is that CARML is incorrectly finding that the logical sources have different sources. This is due to a bug in determining the equality of the xml namespace declarations in the mapping.
So because the namespaces are declared multiples times with different blank node ids, CARML is (incorrectly) finding them to not be equal. 

Further on in the process this causes conflicts because the same input is then read multiple times, instead of once. Sometimes this still works, by chance, but sometimes it doesn't.

I'll try and roll out a fix ASAP. In the meantime, updating your mapping to only use a single instance of the "stdin" source should also resolve this issue.

**Comment by pmaria** [comment]
@mchlrch released 0.4.9 with the fix. Artifacts should be available soon in Maven Central.

**Comment by mchlrch** [comment]
> @mchlrch released 0.4.9 with the fix. Artifacts should be available soon in Maven Central.

Thanks a lot! 👍
